File: tool/import_kanjivg.py
#!/usr/bin/env python3
"""Download KanjiVG and emit stroke JSON for kanji in this app's dataset.

Re-run this script to refresh stroke data after a KanjiVG release. Application
code does not need to change when the dataset is updated.

Usage (from the repository root):

    python tool/import_kanjivg.py
"""

from __future__ import annotations

import json
import re
import sys
import urllib.request
import zipfile
from pathlib import Path
from xml.etree import ElementTree as ET

KANJIVG_RELEASE = "r20250816"
KANJIVG_ZIP_NAME = "kanjivg-20250816-main.zip"
KANJIVG_ZIP_URL = (
    "https://github.com/KanjiVG/kanjivg/releases/download/"
    f"{KANJIVG_RELEASE}/{KANJIVG_ZIP_NAME}"
)
KANJIVG_COPYING_URL = (
    "https://raw.githubusercontent.com/KanjiVG/kanjivg/master/COPYING"
)
KANJIVG_SITE = "https://kanjivg.tagaini.net"
KANJIVG_LICENSE = "Creative Commons Attribution-Share Alike 3.0"

SVG_NS = "http://www.w3.org/2000/svg"
PATH_TAG = f"{{{SVG_NS}}}path"
TEXT_TAG = f"{{{SVG_NS}}}text"
SVG_TAG = f"{{{SVG_NS}}}svg"
G_TAG = f"{{{SVG_NS}}}g"

CHARACTER_RE = re.compile(r"character:\s*'(.+?)'")
DOCTYPE_RE = re.compile(r"<!DOCTYPE[\s\S]*?\]>\s*", re.MULTILINE)
STROKE_ID_RE = re.compile(r"-s(\d+)$")
MATRIX_RE = re.compile(
    r"matrix\(\s*([-\d.eE]+)\s+([-\d.eE]+)\s+([-\d.eE]+)\s+"
    r"([-\d.eE]+)\s+([-\d.eE]+)\s+([-\d.eE]+)\s*\)"
)
TRANSLATE_RE = re.compile(
    r"translate\(\s*([-\d.eE]+)(?:\s*[,\s]\s*([-\d.eE]+))?\s*\)"
)


def repo_root() -> Path:
    return Path(__file__).resolve().parents[1]


def unicode_hex(character: str) -> str:
    return f"{ord(character):05x}"


def read_dataset_characters(data_path: Path) -> list[str]:
    text = data_path.read_text(encoding="utf-8")
    characters: list[str] = []
    seen: set[str] = set()
    for match in CHARACTER_RE.finditer(text):
        glyph = match.group(1)
        if glyph in seen:
            continue
        seen.add(glyph)
        characters.append(glyph)
    return characters


def download(url: str, dest: Path) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    print(f"Downloading {url}")
    request = urllib.request.Request(url, headers={"User-Agent": "5-Minute-Kanji"})
    with urllib.request.urlopen(request) as response:
        dest.write_bytes(response.read())
    print(f"Saved {dest} ({dest.stat().st_size} bytes)")


def strip_doctype(svg_text: str) -> str:
    return DOCTYPE_RE.sub("", svg_text, count=1)


def parse_viewbox(root: ET.Element) -> list[float]:
    raw = root.get("viewBox") or root.get("viewbox") or "0 0 109 109"
    parts = [float(part) for part in raw.replace(",", " ").split() if part]
    if len(parts) != 4:
        return [0.0, 0.0, 109.0, 109.0]
    return parts


def stroke_sort_key(path: ET.Element) -> int:
    match = STROKE_ID_RE.search(path.get("id") or "")
    if match:
        return int(match.group(1))
    return 10_000


def parse_number_position(transform: str | None) -> list[float] | None:
    if not transform:
        return None
    matrix = MATRIX_RE.search(transform)
    if matrix:
        return [round(float(matrix.group(5)), 2), round(float(matrix.group(6)), 2)]
    translate = TRANSLATE_RE.search(transform)
    if translate:
        x = round(float(translate.group(1)), 2)
        y = round(float(translate.group(2) or 0.0), 2)
        return [x, y]
    return None


def extract_stroke_data(svg_text: str, character: str) -> dict:
    xml = strip_doctype(svg_text)
    root = ET.fromstring(xml)
    if root.tag != SVG_TAG:
        svg = root.find(f".//{SVG_TAG}")
        if svg is None:
            raise ValueError("SVG root not found")
        root = svg

    hex_id = unicode_hex(character)
    stroke_paths_group = None
    stroke_numbers_group = None
    for group in root.iter(G_TAG):
        group_id = group.get("id") or ""
        if group_id.startswith("kvg:StrokePaths"):
            stroke_paths_group = group
        elif group_id.startswith("kvg:StrokeNumbers"):
            stroke_numbers_group = group

    search_root = stroke_paths_group if stroke_paths_group is not None else root
    paths = [node for node in search_root.iter(PATH_TAG) if node.get("d")]
    paths.sort(key=stroke_sort_key)
    if not paths:
        raise ValueError("no stroke paths")

    numbers: dict[int, list[float]] = {}
    if stroke_numbers_group is not None:
        for text_node in stroke_numbers_group.iter(TEXT_TAG):
            label = "".join(text_node.itertext()).strip()
            if not label.isdigit():
                continue
            position = parse_number_position(text_node.get("transform"))
            if position is None:
                continue
            numbers[int(label)] = position

    strokes = []
    for index, path in enumerate(paths, start=1):
        entry: dict = {"d": path.get("d", "").strip()}
        if index in numbers:
            entry["number"] = numbers[index]
        strokes.append(entry)

    return {
        "character": character,
        "unicode": hex_id,
        "viewBox": parse_viewbox(root),
        "strokes": strokes,
    }


def find_svg_in_zip(archive: zipfile.ZipFile, hex_id: str) -> str | None:
    suffix = f"/{hex_id}.svg"
    exact = f"{hex_id}.svg"
    for name in archive.namelist():
        normalized = name.replace("\\", "/")
        if normalized.endswith(suffix) or normalized == exact:
            if "/" in normalized and not normalized.rsplit("/", 1)[0].endswith("kanji"):
                continue
            return name
    for name in archive.namelist():
        if name.replace("\\", "/").endswith(f"{hex_id}.svg"):
            return name
    return None


def write_source_markdown(
    path: Path,
    *,
    imported: int,
    missing: list[str],
) -> None:
    missing_lines = (
        "\n".join(f"- {glyph} (U+{ord(glyph):04X})" for glyph in missing)
        if missing
        else "None."
    )
    path.write_text(
        f"""# KanjiVG stroke data

Stroke-order assets in `assets/kanji/strokes/` are derived from [KanjiVG]({KANJIVG_SITE}).

| Field | Value |
| --- | --- |
| Source | KanjiVG |
| Release | `{KANJIVG_RELEASE}` |
| Archive | `{KANJIVG_ZIP_NAME}` |
| Download | {KANJIVG_ZIP_URL} |
| License | {KANJIVG_LICENSE} |
| Copyright | © 2009–2026 Ulrich Apel |
| Imported kanji | {imported} |
| Missing kanji | {len(missing)} |

The JSON files are an adaptation of the KanjiVG SVG stroke paths for the kanji
currently included in this app. They are **not** the full KanjiVG repository.

To refresh the data after a new KanjiVG release, update the constants at the
top of `tool/import_kanjivg.py` if needed and run:

```
python tool/import_kanjivg.py
```

Missing from this KanjiVG release:

{missing_lines}

See `third_party/kanjivg/` for the license text and attribution notice.
""",
        encoding="utf-8",
    )


def write_notice(path: Path) -> None:
    path.write_text(
        f"""KanjiVG
Copyright © 2009-2026 Ulrich Apel

This product includes stroke-order data derived from KanjiVG
({KANJIVG_SITE}), released under the {KANJIVG_LICENSE} license.

The JSON files in assets/kanji/strokes/ are an adaptation of KanjiVG SVG
data and are therefore also licensed under CC BY-SA 3.0.

Full license text: COPYING
""",
        encoding="utf-8",
    )


def main() -> int:
    root = repo_root()
    data_path = root / "lib" / "data" / "hardcoded_kanji_data.dart"
    output_dir = root / "assets" / "kanji" / "strokes"
    source_md = root / "assets" / "kanji" / "SOURCE.md"
    third_party = root / "third_party" / "kanjivg"
    cache_dir = root / "tool" / ".cache"
    zip_path = cache_dir / KANJIVG_ZIP_NAME
    copying_path = third_party / "COPYING"

    characters = read_dataset_characters(data_path)
    if not characters:
        print("No kanji characters found in hardcoded_kanji_data.dart", file=sys.stderr)
        return 1
    print(f"Dataset contains {len(characters)} unique kanji")

    if not zip_path.exists():
        download(KANJIVG_ZIP_URL, zip_path)
    else:
        print(f"Using cached archive {zip_path}")

    if not copying_path.exists():
        download(KANJIVG_COPYING_URL, copying_path)
    write_notice(third_party / "NOTICE")

    output_dir.mkdir(parents=True, exist_ok=True)
    for stale in output_dir.glob("*.json"):
        stale.unlink()

    imported = 0
    missing: list[str] = []
    with zipfile.ZipFile(zip_path) as archive:
        for glyph in characters:
            hex_id = unicode_hex(glyph)
            member = find_svg_in_zip(archive, hex_id)
            if member is None:
                print(f"MISSING  {glyph}  {hex_id}.svg")
                missing.append(glyph)
                continue
            svg_text = archive.read(member).decode("utf-8")
            try:
                payload = extract_stroke_data(svg_text, glyph)
            except Exception as error:  # noqa: BLE001 - report and continue
                print(f"FAILED   {glyph}  {hex_id}.svg  ({error})")
                missing.append(glyph)
                continue
            dest = output_dir / f"{hex_id}.json"
            dest.write_text(
                json.dumps(payload, ensure_ascii=False, separators=(",", ":")),
                encoding="utf-8",
            )
            imported += 1

    write_source_markdown(source_md, imported=imported, missing=missing)
    print(f"Wrote {imported} stroke files to {output_dir}")
    if missing:
        print(f"{len(missing)} kanji missing from KanjiVG:")
        for glyph in missing:
            print(f"  {glyph} U+{ord(glyph):04X}")
    else:
        print("All dataset kanji have KanjiVG stroke data.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tool/import_kanjivg.dart
// Downloads KanjiVG and writes stroke JSON for kanji in this app's dataset.
//
// Re-run after a KanjiVG release. Application code does not need to change.
//
//   dart run tool/import_kanjivg.dart

import 'dart:convert';
import 'dart:io';

const kanjiVgRelease = 'r20250816';
const kanjiVgZipName = 'kanjivg-20250816-main.zip';
const kanjiVgZipUrl =
    'https://github.com/KanjiVG/kanjivg/releases/download/$kanjiVgRelease/$kanjiVgZipName';
const kanjiVgCopyingUrl =
    'https://raw.githubusercontent.com/KanjiVG/kanjivg/master/COPYING';
const kanjiVgSite = 'https://kanjivg.tagaini.net';
const kanjiVgLicense = 'Creative Commons Attribution-Share Alike 3.0';

final characterPattern = RegExp(r"character:\s*'(.+?)'");
final strokeIdPattern = RegExp(r'-s(\d+)$');
final matrixPattern = RegExp(
  r'matrix\(\s*([-\d.eE]+)\s+([-\d.eE]+)\s+([-\d.eE]+)\s+([-\d.eE]+)\s+([-\d.eE]+)\s+([-\d.eE]+)\s*\)',
);
final translatePattern = RegExp(
  r'translate\(\s*([-\d.eE]+)(?:\s*[,\s]\s*([-\d.eE]+))?\s*\)',
);
final pathPattern = RegExp(r'<path\b([^>]*)/?>');
final textPattern = RegExp(r'<text\b([^>]*)>([^<]*)</text>');
final attributePattern = RegExp(r'''([:\w]+)\s*=\s*("([^"]*)"|'([^']*)')''');
final viewBoxPattern = RegExp(r'''viewBox\s*=\s*("([^"]*)"|'([^']*)')''');

Future<void> main() async {
  final dataFile = File('lib/data/hardcoded_kanji_data.dart');
  if (!dataFile.existsSync()) {
    stderr.writeln('Run this script from the repository root.');
    exitCode = 1;
    return;
  }

  final characters = readDatasetCharacters(await dataFile.readAsString());
  stdout.writeln('Dataset contains ${characters.length} unique kanji');

  final cacheDir = Directory('tool/.cache');
  final zipFile = File('${cacheDir.path}/$kanjiVgZipName');
  if (!zipFile.existsSync()) {
    await download(kanjiVgZipUrl, zipFile);
  } else {
    stdout.writeln('Using cached archive ${zipFile.path}');
  }

  final thirdParty = Directory('third_party/kanjivg');
  final copyingFile = File('${thirdParty.path}/COPYING');
  if (!copyingFile.existsSync()) {
    await download(kanjiVgCopyingUrl, copyingFile);
  }
  await File('${thirdParty.path}/NOTICE').writeAsString(noticeText());

  final extractDir = Directory('${cacheDir.path}/kanjivg-extract');
  if (extractDir.existsSync()) {
    extractDir.deleteSync(recursive: true);
  }
  extractDir.createSync(recursive: true);
  final extract = await Process.run('tar', [
    '-xf',
    zipFile.path,
    '-C',
    extractDir.path,
  ]);
  if (extract.exitCode != 0) {
    stderr.writeln(extract.stderr);
    stderr.writeln('Failed to extract ${zipFile.path}');
    exitCode = 1;
    return;
  }

  final svgByHex = <String, File>{};
  await for (final entity in extractDir.list(recursive: true)) {
    if (entity is! File || !entity.path.toLowerCase().endsWith('.svg')) {
      continue;
    }
    final name = entity.uri.pathSegments.last;
    final hex = name.substring(0, name.length - 4).toLowerCase();
    final parent = entity.parent.path.replaceAll('\\', '/');
    if (parent.endsWith('/kanji') || !svgByHex.containsKey(hex)) {
      svgByHex[hex] = entity;
    }
  }

  final outputDir = Directory('assets/kanji/strokes');
  if (outputDir.existsSync()) {
    for (final file in outputDir.listSync()) {
      if (file is File && file.path.endsWith('.json')) {
        file.deleteSync();
      }
    }
  } else {
    outputDir.createSync(recursive: true);
  }

  var imported = 0;
  final missing = <String>[];
  for (final glyph in characters) {
    final hex = unicodeHex(glyph);
    final svgFile = svgByHex[hex];
    if (svgFile == null) {
      stdout.writeln('MISSING  $glyph  $hex.svg');
      missing.add(glyph);
      continue;
    }
    try {
      final payload = extractStrokeData(svgFile.readAsStringSync(), glyph);
      File(
        '${outputDir.path}/$hex.json',
      ).writeAsStringSync(jsonEncode(payload));
      imported++;
    } catch (error) {
      stdout.writeln('FAILED   $glyph  $hex.svg  ($error)');
      missing.add(glyph);
    }
  }

  await File(
    'assets/kanji/SOURCE.md',
  ).writeAsString(sourceMarkdown(imported: imported, missing: missing));
  stdout.writeln('Wrote $imported stroke files to ${outputDir.path}');
  if (missing.isEmpty) {
    stdout.writeln('All dataset kanji have KanjiVG stroke data.');
  } else {
    stdout.writeln('${missing.length} kanji missing from KanjiVG:');
    for (final glyph in missing) {
      final code = glyph.runes.first
          .toRadixString(16)
          .toUpperCase()
          .padLeft(4, '0');
      stdout.writeln('  $glyph U+$code');
    }
  }
}

List<String> readDatasetCharacters(String source) {
  final characters = <String>[];
  final seen = <String>{};
  for (final match in characterPattern.allMatches(source)) {
    final glyph = match.group(1)!;
    if (seen.add(glyph)) {
      characters.add(glyph);
    }
  }
  return characters;
}

String unicodeHex(String character) {
  return character.runes.first.toRadixString(16).padLeft(5, '0');
}

Future<void> download(String url, File dest) async {
  dest.parent.createSync(recursive: true);
  stdout.writeln('Downloading $url');
  final client = HttpClient();
  try {
    final request = await client.getUrl(Uri.parse(url));
    request.followRedirects = true;
    final response = await request.close();
    if (response.statusCode >= 400) {
      throw HttpException('HTTP ${response.statusCode} for $url');
    }
    await response.pipe(dest.openWrite());
    stdout.writeln('Saved ${dest.path} (${dest.lengthSync()} bytes)');
  } finally {
    client.close();
  }
}

Map<String, String> parseAttributes(String raw) {
  final attributes = <String, String>{};
  for (final match in attributePattern.allMatches(raw)) {
    attributes[match.group(1)!] = match.group(3) ?? match.group(4) ?? '';
  }
  return attributes;
}

List<double>? parseNumberPosition(String? transform) {
  if (transform == null || transform.isEmpty) return null;
  final matrix = matrixPattern.firstMatch(transform);
  if (matrix != null) {
    return [double.parse(matrix.group(5)!), double.parse(matrix.group(6)!)];
  }
  final translate = translatePattern.firstMatch(transform);
  if (translate != null) {
    return [
      double.parse(translate.group(1)!),
      double.parse(translate.group(2) ?? '0'),
    ];
  }
  return null;
}

Map<String, Object> extractStrokeData(String svgText, String character) {
  final hexId = unicodeHex(character);
  final pathsStart = svgText.indexOf('kvg:StrokePaths');
  final numbersStart = svgText.indexOf('kvg:StrokeNumbers');
  final pathsSection = pathsStart >= 0
      ? svgText.substring(pathsStart, numbersStart >= 0 ? numbersStart : null)
      : svgText;

  final indexed = <({int index, String d})>[];
  for (final match in pathPattern.allMatches(pathsSection)) {
    final attributes = parseAttributes(match.group(1)!);
    final d = attributes['d'];
    if (d == null || d.isEmpty) continue;
    final id = attributes['id'] ?? '';
    final strokeMatch = strokeIdPattern.firstMatch(id);
    indexed.add((
      index: strokeMatch == null
          ? indexed.length + 1
          : int.parse(strokeMatch.group(1)!),
      d: d,
    ));
  }
  indexed.sort((a, b) => a.index.compareTo(b.index));
  if (indexed.isEmpty) {
    throw StateError('no stroke paths');
  }

  final numbers = <int, List<double>>{};
  if (numbersStart >= 0) {
    for (final match in textPattern.allMatches(
      svgText.substring(numbersStart),
    )) {
      final label = match.group(2)!.trim();
      final index = int.tryParse(label);
      if (index == null) continue;
      final attributes = parseAttributes(match.group(1)!);
      final position = parseNumberPosition(attributes['transform']);
      if (position != null) {
        numbers[index] = [
          double.parse(position[0].toStringAsFixed(2)),
          double.parse(position[1].toStringAsFixed(2)),
        ];
      }
    }
  }

  final viewBoxMatch = viewBoxPattern.firstMatch(svgText);
  final viewBoxRaw =
      viewBoxMatch?.group(2) ?? viewBoxMatch?.group(3) ?? '0 0 109 109';
  final viewBox = [
    for (final part in viewBoxRaw.replaceAll(',', ' ').split(RegExp(r'\s+')))
      if (part.isNotEmpty) double.parse(part),
  ];

  return {
    'character': character,
    'unicode': hexId,
    'viewBox': viewBox.length == 4 ? viewBox : [0.0, 0.0, 109.0, 109.0],
    'strokes': [
      for (final stroke in indexed)
        {
          'd': stroke.d,
          if (numbers[stroke.index] != null) 'number': numbers[stroke.index],
        },
    ],
  };
}

String noticeText() {
  return '''
KanjiVG
Copyright © 2009-2026 Ulrich Apel

This product includes stroke-order data derived from KanjiVG
($kanjiVgSite), released under the $kanjiVgLicense license.

The JSON files in assets/kanji/strokes/ are an adaptation of KanjiVG SVG
data and are therefore also licensed under CC BY-SA 3.0.

Full license text: COPYING
''';
}

String sourceMarkdown({required int imported, required List<String> missing}) {
  final missingLines = missing.isEmpty
      ? 'None.'
      : missing
            .map((glyph) {
              final code = glyph.runes.first
                  .toRadixString(16)
                  .toUpperCase()
                  .padLeft(4, '0');
              return '- $glyph (U+$code)';
            })
            .join('\n');

  return '''
# KanjiVG stroke data

Stroke-order assets in `assets/kanji/strokes/` are derived from [KanjiVG]($kanjiVgSite).

| Field | Value |
| --- | --- |
| Source | KanjiVG |
| Release | `$kanjiVgRelease` |
| Archive | `$kanjiVgZipName` |
| Download | $kanjiVgZipUrl |
| License | $kanjiVgLicense |
| Copyright | © 2009–2026 Ulrich Apel |
| Imported kanji | $imported |
| Missing kanji | ${missing.length} |

The JSON files are an adaptation of the KanjiVG SVG stroke paths for the kanji
currently included in this app. They are **not** the full KanjiVG repository.

To refresh the data after a new KanjiVG release, update the constants at the
top of `tool/import_kanjivg.dart` if needed and run:

```
dart run tool/import_kanjivg.dart
```

A Python equivalent lives at `tool/import_kanjivg.py`.

Missing from this KanjiVG release:

$missingLines

See `third_party/kanjivg/` for the license text and attribution notice.
''';
}
